contracts/truefi2/TrueMultiFarm.sol

```diff
@@ -0,0 +1,376 @@
+        return stakes[token].staked[staker];
+    }
+
+    /**
+     * @dev Initialize staking pool with a Distributor contract
+     * The distributor contract calculates how much TRU rewards this contract
+     * gets, and stores TRU for distribution.
+     * @param _trueDistributor Distributor contract
+     */
+    function initialize(ITrueDistributor _trueDistributor) public initializer {
+        UpgradeableClaimable.initialize(msg.sender);
+        trueDistributor = _trueDistributor;
+        rewardToken = _trueDistributor.trustToken();
+        require(trueDistributor.farm() == address(this), "TrueMultiFarm: Distributor farm is not set");
+    }
+
+    /**
+     * @dev Stake tokens for TRU rewards.
+     * Also claims any existing rewards.
+     * @param amount Amount of tokens to stake
+     */
+    function stake(IERC20 token, uint256 amount) external override hasShares(token) update(token) {
+        if (stakerRewards[token].claimableReward[msg.sender] > 0) {
+            _claim(token);
+        }
+        stakes[token].staked[msg.sender] = stakes[token].staked[msg.sender].add(amount);
+        stakes[token].totalStaked = stakes[token].totalStaked.add(amount);
+
+        token.safeTransferFrom(msg.sender, address(this), amount);
+        emit Stake(token, msg.sender, amount);
+    }
+
+    /**
+     * @dev Remove staked tokens
+     * @param amount Amount of tokens to unstake
+     */
+    function unstake(IERC20 token, uint256 amount) external override update(token) {
+        _unstake(token, amount);
+    }
+
+    /**
+     * @dev Claim TRU rewards
+     */
+    function claim(IERC20[] calldata tokens) external override {
+        distribute();
+        for (uint256 i = 0; i < tokens.length; i++) {
+            updateRewards(tokens[i]);
+        }
+        for (uint256 i = 0; i < tokens.length; i++) {
+            _claim(tokens[i]);
+        }
+    }
+
+    /**
+     * @dev Unstake amount and claim rewards
+     */
+    function exit(IERC20[] calldata tokens) external override {
+        distribute();
+        for (uint256 i = 0; i < tokens.length; i++) {
+            updateRewards(tokens[i]);
+        }
+        for (uint256 i = 0; i < tokens.length; i++) {
+            _unstake(tokens[i], stakes[tokens[i]].staked[msg.sender]);
+            _claim(tokens[i]);
+        }
+    }
+
+    /**
+     * @dev Set shares for farms
+     * Example: setShares([DAI, USDC], [1, 2]) will ensure that 33.(3)% of rewards will go to DAI farm and rest to USDC farm
+     * If later setShares([DAI, TUSD], [2, 1]) will be called then shares of DAI will grow to 2, shares of USDC won't change and shares of TUSD will be 1
+     * So this will give 40% of rewards going to DAI farm, 40% to USDC and 20% to TUSD
+     * @param tokens Token addresses
+     * @param updatedShares share of the i-th token in the multifarm
+     */
+    function setShares(IERC20[] calldata tokens, uint256[] calldata updatedShares) public onlyOwner {
```

Comment:
Would it be better to do something more similar to MasterChef (https://github.com/sushiswap/sushiswap/blob/master/contracts/MasterChef.sol)

In MasterChef, there is simply a total allocation points, and each farm has a certain number of points. Adding points to a farm increases the number of total allocation points, removing points from a farm decreases the total allocation points. Masterchef then has a total block reward which is split based on farm_points / total_points. This might simplify this system unless you think setting shares in a single call is more efficient.

Reply:
Looks like Masterchef does it similarly to us https://github.com/sushiswap/sushiswap/blob/master/contracts/MasterChef.sol#L106
We only added an option of changing multiple allocation points or shares in our case. It doesn't complicate the code by much and can be a handy feature considering we have Governance and don't want to have lots of polls to rebalance the farm